clang/utils/TableGen/ClangDiagnosticsEmitter.cpp

```diff
@@ -255,20 +257,18 @@ class InferPedantic {
       GMap;
 
   DiagGroupParentMap &DiagGroupParents;
-  const std::vector<Record*> &Diags;
-  const std::vector<Record*> DiagGroups;
+  ArrayRef<const Record *> Diags;
```

Comment:
Are we sure this isn't going to cause lifetime issues because `ArrayRef` does not own its memory?

Reply:
Yes. If you look at https://discourse.llvm.org/t/psa-planned-changes-to-tablegen-getallderiveddefinitions-api-potential-downstream-breakages/81089

one of the changes in that RecordKeeper::getAllDerivedDefinitions() returns an ArrayRef to an internal cached vector, so this memory is owned by RecordKeeper, which for all TableGen backends is guaranteed to be live.

Here, Diags is set using  getAllDerivedDefinitions(), so should be safe. 

Note that previous code also uses a reference to a std::vector for Diags. However, in the earlier version getAllDerivedDefinitions() returned a vector by value. So, this is a reference to an unnamed temporary. I think C++ requires that the compiler extends the lifetime of the unnamed temp is it sees a reference to it being used, I am assuming that is correct even in the older version when we had the reference pointing to an unnamed temp return value of getAllDerivedDefinitions().

Reply:
Let me know if that answers your question. This PR will unblock some other PR's of similar nature that depend on this one.